docker/Dockerfile

```diff
@@ -1,58 +1,129 @@
-RUN pip3 install --upgrade pip torch==${TORCH_VERSION}+cu111 torchvision==${TORCHVISION_VERSION}+cu111 -f https://download.pytorch.org/whl/torch_stable.html
+RUN $VENV/bin/pip install --upgrade pip torch==${TORCH_VERSION}+cu111 torchvision==${TORCHVISION_VERSION}+cu111 -f https://download.pytorch.org/whl/torch_stable.html
+
+FROM cuda-$CUDA_VERSION as cuda_builder
+ARG VENV
+RUN $VENV/bin/pip install --upgrade setuptools wheel
+ARG BRANCH
+ARG VERSION
+RUN \
+    if [ -z "${VERSION}" ] ; then  \
+    echo VERSION not provided && \
+      if [ -z "${BRANCH}" ]; then  \
+        echo cloning main \
+          && git clone https://github.com/neuralmagic/sparseml.git --depth 1 -b main; \
+      else echo cloning $BRANCH \
+          && git clone https://github.com/neuralmagic/sparseml.git --depth 1 -b $BRANCH; \
+      fi; \
+    else echo VERSION is $VERSION \
+      &&  git clone https://github.com/neuralmagic/sparseml.git --depth 1 -b release/${VERSION}; \
+    fi;
+
+
+FROM cuda_builder AS container_branch_base
+ARG VENV
+ENV PATH="${VENV}/bin:$PATH"
+ENV PIP_DEFAULT_TIMEOUT=200
+ARG SPARSEML_INSTALLABLE=./sparseml
+RUN $VENV/bin/pip install --no-cache-dir "${SPARSEML_INSTALLABLE}"
+
+
+FROM container_branch_base as container_branch_all
+ARG VENV
+ARG SPARSEML_INSTALLABLE=./sparseml[onnxruntime,torchvision,ultralytics]
+RUN $VENV/bin/pip install --no-cache-dir "${SPARSEML_INSTALLABLE}"
+RUN sparseml.transformers.question_answering --help \
+    && sparseml.yolov5.train --help \
+    && sparseml.ultralytics.train --help
+
+FROM container_branch_base as container_branch_transformers
+RUN sparseml.transformers.question_answering --help
+
+
+FROM container_branch_base as container_branch_torchvision
+ARG VENV
+ARG SPARSEML_INSTALLABLE=./sparseml[torchvision]
+RUN $VENV/bin/pip install --no-cache-dir "${SPARSEML_INSTALLABLE}"
+
+FROM container_branch_base as container_branch_ultralytics
+ARG VENV
+ARG SPARSEML_INSTALLABLE=./sparseml[onnxruntime,ultralytics]
+RUN $VENV/bin/pip install --no-cache-dir "${SPARSEML_INSTALLABLE}"
+RUN sparseml.ultralytics.train --help
+
+
+FROM container_branch_base as container_branch_yolov5
+ARG VENV
+ARG SPARSEML_INSTALLABLE=./sparseml
+RUN $VENV/bin/pip install --no-cache-dir "${SPARSEML_INSTALLABLE}"
+RUN sparseml.yolov5.train --help
+
+
+FROM container_branch_base as container_branch_dev
+ARG VENV
+ARG SPARSEML_INSTALLABLE=./sparseml[dev]
+RUN $VENV/bin/pip install -e "${SPARSEML_INSTALLABLE}"
 
-FROM cuda-$CUDA_VERSION as target
 
-RUN pip3 install --upgrade setuptools wheel
+FROM container_branch_${DEPS} AS build
+RUN echo Build complete, going onto prod
 
-ARG GIT_CHECKOUT
-# Install SparseML
-# if $GIT_CHECKOUT is not specified - just install from pypi
-RUN if [ -z "${GIT_CHECKOUT}" ] ; then pip3 install --no-cache-dir --upgrade sparseml[dev,torchvision] ; fi
+FROM base as prod
+ARG VENV
+COPY --from=build $VENV $VENV
+ENV PATH="${VENV}/bin:$PATH"
+HEALTHCHECK CMD python -c 'import sparseml'
```

Comment:
is this healthcheck run on every sub-container?

Reply:
The healthcheck is run once per build, and not for each individual stage.